amaranth/lib/crc/__init__.py

```diff
@@ -0,0 +1,417 @@
+    CRC register will always take on a fixed value known as the residue.  The
+    ``match_detected`` output indicates whether the CRC register currently
+    contains this residue.
+
+    Parameters
+    ----------
+    parameters : Parameters
+        CRC parameters.
+
+    Attributes
+    ----------
+    start : Signal(), in
+        Assert to indicate the start of a CRC computation, re-initialising
+        the CRC register to the initial value. May be asserted simultaneously
+        with ``valid`` or by itself.
+    data : Signal(data_width), in
+        Data word to add to CRC when ``valid`` is asserted.
+    valid : Signal(), in
+        Assert when ``data`` is valid to add the data word to the CRC.
+    crc : Signal(crc_width), out
+        Registered CRC output value, updated one clock cycle after ``valid``
+        becomes asserted.
+    match_detected : Signal(), out
+        Asserted if the current CRC value indicates a valid codeword has been
+        received.
+    """
+    def __init__(self, parameters):
+        if not isinstance(parameters, Parameters):
+            raise TypeError("Parameters must be of type Parameters")
+        self.crc_width = parameters.crc_width
+        self.data_width = parameters.data_width
+        self.polynomial = parameters.polynomial
+        self.initial_crc = Const(parameters.initial_crc, self.crc_width)
+        self.reflect_input = parameters.reflect_input
+        self.reflect_output = parameters.reflect_output
+        self.xor_output = parameters.xor_output
+        self._matrix_f, self._matrix_g = parameters._matrices()
+        self._residue = parameters.residue()
+
+        self.start = Signal()
+        self.data = Signal(self.data_width)
+        self.valid = Signal()
+        self.crc = Signal(self.crc_width)
+        self.match_detected = Signal()
+
+    def elaborate(self, platform):
+        m = Module()
+
+        crc_reg = Signal(self.crc_width, reset=self.initial_crc.value)
+        data_in = Signal(self.data_width)
+
+        # Optionally bit-reflect input words.
+        if self.reflect_input:
+            m.d.comb += data_in.eq(self.data[::-1])
+        else:
+            m.d.comb += data_in.eq(self.data)
+
+        # Optionally bit-reflect and then XOR the output.
+        if self.reflect_output:
+            m.d.comb += self.crc.eq(crc_reg[::-1] ^ self.xor_output)
+        else:
+            m.d.comb += self.crc.eq(crc_reg ^ self.xor_output)
+
+        # Compute next CRC state.
+        source = Mux(self.start, self.initial_crc, crc_reg)
+        with m.If(self.valid):
+            for i in range(self.crc_width):
+                bit = 0
+                for j in range(self.crc_width):
+                    if self._matrix_f[j][i]:
+                        bit ^= source[j]
+                for j in range(self.data_width):
+                    if self._matrix_g[j][i]:
+                        bit ^= data_in[j]
+                m.d.sync += crc_reg[i].eq(bit)
```

Comment:
The inner loops can be removed if the matrices use bit-vectors instead.
```python
m.d.sync += crc_reg.bit_select(i, 1).eq(
	(source & self._matrix_f[i]).xor() ^ (data_in & self._matrix_g[i]).xor()
)
```
This also allows to skip the bit conversion during the matrix generation step.
However this might reduce the readability of the algorithm.

Reply:
Could you explain a bit more what the impact of this chanage is besides readability?

Reply:
It gives the synthesis a bit more to chew on but I think optimise to the same thing, i.e., previously:

    crc_reg[3] = 0 ^ source[2] ^ source[4] ^ data_in[3] ^ data_in[5]

instead it would be:

    crc_reg[3] = (source[0] & 0) ^ (source[1] ^ 0) & (source[2] ^ 1) ^ (source[3] & 0) ^ (source[4] & 1) ^ (source[5] & 0) ^ (data_in[0] & 0) ^ (data_in[1] & 0) & (data_in[2] & 0) ^ (data_in[3] & 1) ^ (data_in[4] & 0) & (data_in[5] & 1)

I don't remember if `.xor()` existed as a method when I wrote this the first time or I might have used it.

Reply:
I just uploaded my parallel CRC implementation that uses this approach.
It still needs a bit more documentation, but maybe it can help as a reference.

https://gist.github.com/hannesha/3d8adf48fc85013e8c77d42492611864

Reply:
I don't think there's any need to change the current implementation, right? The results should be the same, we just do the loop over the bits in the CRC generator instead of later on in the stack. I like that the current way is clear that each new bit is only generated from a small subset of the state/input bits, but I think it's marginal either way.

Reply:
Yes the current implementation is fine. Compared to my approach it only generates the required bits without relying on constant folding to produce the same logic.